### DISCUSSION TITLE
How to override a native function

### DISCUSSION THREAD

**Discussion by yukulele** [violentmonkey/violentmonkey#1655] [Q&A]
I want to override `window.open()`:

```javascript
const _open = window.open.bind(window)

window.open = (url, ...args) => {
  if(!confirm(`open "${url}" ?`) return
  _open(url, ...args)
}
```

This code works when I type it in the console but doesn't work from an userscript: it keep `window.open()` as the native function.

Is there a way to override `window` properties?

**Comment by tophf** [comment]
If your script has `@grant` that is not `none` you need to use `unsafeWindow` instead of `window`.

**Reply by yukulele** [reply]
I tryed this but it doesn't work:
```javascript
// ==UserScript==
// @name        alert
// @namespace   gitub alert test
// @match       https://github.com/*
// @grant       unsafeWindow
// @version     1.0
// @author      -
// @description 09/11/2022 13:01:32
// ==/UserScript==

const __alert = unsafeWindow.alert.bind(unsafeWindow)
unsafeWindow.alert = message => __alert('[ALERT]: ' + message)

unsafeWindow.alert('foo') // (A)
```
(A) correctly displays an alert popup with `[ALERT]: foo` but `alert('bar')` in console displays an alert popup with `bar`.

**Reply by tophf** [reply]
There's no need for `@grant` at all in this case. I guess you're using Firefox so you'll have to either disable CSP on github (or in the browser globally, look for instructions on doing that) or use wrappedJSObject + exportFunction:

```js
const {alert} = wrappedJSObject;
wrappedJSObject.alert = exportFunction(msg => alert('[ALERT]: ' + msg), wrappedJSObject);
```
